[TS migration] Migrate 'ReimbursementAccount' to TypeScript

mateuuszzzzz opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/24928
PROPOSAL:


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Go to settings of a workspace that you own (or have admin permissions).
- [ ] Go to Reimbursements section and press `Connect bank account`.
- [ ] Test both options for adding account (plaid and manual).
- [ ] Verify that all steps are present in setup flow.
- [ ] Verify that going backward and forward preserves entered data.
- [ ] Verify if errors appear when incorrect data is entered and if those errors are relevant.
- [ ] Check if canceling (`Start over` button) the flow at any point does not cause problems with future bank account setups.(e.g. if old errors don't appear in future setups and information related to previous setup is removed). 

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

- 

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

Same as tests

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
Not sure if there are any videos to include since the pr modifies action code.
<details>
<summary>Android: Native</summary>


https://github.com/Expensify/App/assets/60450261/882787d3-3ac0-44ba-91d7-69ec89e2f089



</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/Expensify/App/assets/60450261/6f925111-df4a-42b7-8c1f-f08f0f9b33e5



</details>

<details>
<summary>iOS: Native</summary>


https://github.com/Expensify/App/assets/60450261/eefd3534-ba2d-48c0-81c2-9a02df86a254



</details>

<details>
<summary>iOS: mWeb Safari</summary>


https://github.com/Expensify/App/assets/60450261/68daa06b-8bec-4911-ad74-954cee4bca46



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/60450261/25b8c583-4410-47a7-860e-89fb379ac8c7



</details>

<details>
<summary>MacOS: Desktop</summary>


https://github.com/Expensify/App/assets/60450261/8114d3a9-33df-418b-836d-04945fa42eba



</details>

mateuuszzzzz committed:
Migrate ReimbursementAccount to ts

mateuuszzzzz committed:
Remove old js implementations

mateuuszzzzz committed:
Add ErrorFields to PersonalBankAccount type

mateuuszzzzz committed:
Adjust types

mateuuszzzzz committed:
Resolve conflicts with main

mateuuszzzzz committed:
Fix typescript errors

mateuuszzzzz commented:
I noticed this typescript error:

`Error: src/pages/ReimbursementAccount/VerifyIdentity/VerifyIdentity.tsx(45,59): error TS2353: Object literal may only specify known properties, and 'isOnfidoSetupComplete' does not exist in type 'Partial<ReimbursementAccountForm>'`

Looks like we missed `isOnfidoSetupComplete`. I wonder which place would be valid for it since `ReimbursementAccountForm` is composition of a few types 🤔

mateuuszzzzz committed:
Fix comment typo

blazejkustra commented on src/libs/actions/ReimbursementAccount/store.ts:
```diff
@@ -46,7 +48,11 @@ function getBankAccountList() {
+        return false;
+    }
+
+    Object.entries(bankAccountList).some(([, bankAccountJSON]) => {
```
Why not Object.values here?

mateuuszzzzz replied:
Ah yeah, honestly I have no idea 😅 But in fact, values seem a better choice here.

blazejkustra reviewed (approved):
One minor comments, otherwise LGTM!

mateuuszzzzz committed:
Change Object.entries to Object.values

mateuuszzzzz committed:
Add missing isOnfidoSetupComplete

VickyStash commented on src/libs/actions/ReimbursementAccount/index.ts:
```diff
@@ -11,26 +13,20 @@ export {setBankAccountFormValidationErrors, setPersonalBankAccountFormValidation
- * @returns {Promise<void>}
  */
-function setBankAccountSubStep(subStep) {
+function setBankAccountSubStep(subStep: BankAccountSubStep | null) {
```
Let's add return type

VickyStash commented on src/libs/actions/ReimbursementAccount/navigation.ts:
```diff
@@ -2,23 +2,22 @@ import Onyx from 'react-native-onyx';
- * @param {string} policyID - The policy ID associated with the bank account.
- * @param {string} [backTo=''] - An optional return path. If provided, it will be URL-encoded and appended to the resulting URL.
+ * @param policyID - The policy ID associated with the bank account.
+ * @param [backTo=''] - An optional return path. If provided, it will be URL-encoded and appended to the resulting URL.
```
```suggestion
 * @param [backTo] - An optional return path. If provided, it will be URL-encoded and appended to the resulting URL.
```

VickyStash commented on src/libs/actions/ReimbursementAccount/store.ts:
```diff
@@ -1,39 +1,41 @@
 
 /** Reimbursement account actively being set up */
-let reimbursementAccountInSetup = {};
+let reimbursementAccountInSetup: ACHData | EmptyObject = {};
```
Maybe we can do the same you did with `credentials` in this file, wdyt? But also I haven't found this values are really being used. It's used only in a function which is used nowhere...
```suggestion
let reimbursementAccountInSetup: OnyxEntry<ACHData> = null;
```

mateuuszzzzz replied:
Hmm but now I think about it and I'm not sure if we use `OnyxEntry` for keys of `Onyx` types.
I mean that `ACHData` is not `OnyxEntry` itself but it is a key of `ReimbursementAccount` 🤔

VickyStash replied:
I'm okay to leave it as is then 👌

VickyStash commented on src/libs/actions/ReimbursementAccount/store.ts:
```diff
@@ -1,39 +1,41 @@
     key: ONYXKEYS.REIMBURSEMENT_ACCOUNT,
     callback: (val) => {
-        reimbursementAccountInSetup = lodashGet(val, 'achData', {});
+        reimbursementAccountInSetup = val?.achData ?? {};
```
```suggestion
        reimbursementAccountInSetup = val?.achData ?? null;
```

kubabutkiewicz commented on src/libs/actions/ReimbursementAccount/index.ts:
```diff
@@ -11,26 +13,20 @@ export {setBankAccountFormValidationErrors, setPersonalBankAccountFormValidation
 }
 
-function setWorkspaceIDForReimbursementAccount(workspaceID) {
+function setWorkspaceIDForReimbursementAccount(workspaceID: string | null) {
```
I search for a usage for this function and it seems its not used anywhere, maybe we can delete it?

mateuuszzzzz replied:
@kubabutkiewicz  there are even more functions that are not used anywhere or are only re-exported in other modules but again not used. I'm not sure if this is a mistake and they should be deleted or temporarily those functions are not in use. cc: @VickyStash 

Here's the full list of what I found:

Not used anywhere:

`getReimbursementAccountInSetup`
`getReimbursementAccountWorkspaceID`
`getCredentials`

Exported in `ReimbursementAccount` (`index.ts`) but not used anywhere:

`setPersonalBankAccountFormValidationErrorFields`
`showBankAccountFormValidationError`
`deleteFromBankAccountList`

Exported in `BankAccount.ts` and `ReimbursementAccount` (`index.ts`) but not used anywhere:

`setBankAccountFormValidationErrors`
`setWorkspaceIDForReimbursementAccount`

kubabutkiewicz replied:
@fabioh8010 do you think its safe to remove them?

fabioh8010 replied:
I think we could check if he had recent activity on these functions, because sometimes we have code that looks unused but it was created with the purpose to be used later (during a feature development for example).

mateuuszzzzz replied:
Well, those functions seems to be pretty old:

`getReimbursementAccountInSetup` (~2 years) https://github.com/Expensify/App/pull/6230
`getReimbursementAccountWorkspaceID` (~2 years) https://github.com/Expensify/App/pull/6230
`getCredentials` (~2 years) https://github.com/Expensify/App/pull/6230
`setPersonalBankAccountFormValidationErrorFields` (~20 months) https://github.com/Expensify/App/pull/9423
`showBankAccountFormValidationError` (~2 years) https://github.com/Expensify/App/pull/6230
`deleteFromBankAccountList` (~2 years) https://github.com/Expensify/App/pull/7557
`setWorkspaceIDForReimbursementAccount` (~2 years) https://github.com/Expensify/App/pull/6230

fabioh8010 replied:
Thanks for the info @mateuuszzzzz ! Well, I suppose we can remove them, let's do it and see how it goes during C+/Exfy reviews.

mateuuszzzzz replied:
Cool, I removed them

mateuuszzzzz committed:
Add return type and change comment

mateuuszzzzz committed:
Remove unused functions

kubabutkiewicz commented on src/libs/actions/ReimbursementAccount/store.ts:
```diff
@@ -0,0 +1,30 @@
+    },
+});
+
+function getBankAccountList() {
```
add return type

kubabutkiewicz commented on src/libs/actions/ReimbursementAccount/store.ts:
```diff
@@ -0,0 +1,30 @@
+    return bankAccountList;
+}
+
+function hasCreditBankAccount() {
```
add return type

kubabutkiewicz commented on src/libs/actions/ReimbursementAccount/store.ts:
```diff
@@ -0,0 +1,30 @@
+        return false;
+    }
+
+    Object.values(bankAccountList).some((bankAccountJSON) => {
```
I think you are missing here `return`

mateuuszzzzz replied:
Good catch!

mateuuszzzzz committed:
Add return types and missing return

blazejkustra commented:
@mateuuszzzzz I agree with Aimane, let's fill out tests and QA sections

aimane-chnaif reviewed (commented):
I see that lots of unused functions are removed but was it discussed?

aimane-chnaif reviewed (approved):
Code looks good

mountiny commented:
@mateuuszzzzz or @blazejkustra can someone merge main please

mateuuszzzzz committed:
Merge branch 'main' into ts/migrate-reimbursement-account

mateuuszzzzz committed:
Merge branch 'main' into ts/migrate-reimbursement-account

mateuuszzzzz commented:
@mountiny @aimane-chnaif done

mateuuszzzzz commented:
> I see that lots of unused functions are removed but was it discussed?

According to this. We had discussion about deleting them in this PR and decided to do it since those functions are pretty old

mountiny commented on src/libs/actions/ReimbursementAccount/errors.ts:
```diff
@@ -0,0 +1,25 @@
+
+/**
+ * Set the current fields with errors.
+
```
```suggestion
```

mountiny commented on src/libs/actions/ReimbursementAccount/resetFreePlanBankAccount.ts:
```diff
@@ -0,0 +1,126 @@
+                {
+                    onyxMethod: Onyx.METHOD.SET,
+                    key: ONYXKEYS.FORMS.REIMBURSEMENT_ACCOUNT_FORM_DRAFT,
+                    value: {
```
In the js file we just set this to empty object, why this change

mateuuszzzzz replied:
Similarly here `typecheck` throws error, because keys of the type used here are not optional

mountiny commented on src/pages/ReimbursementAccount/BankInfo/BankInfo.tsx:
```diff
@@ -118,7 +118,7 @@ function BankInfo({reimbursementAccount, reimbursementAccountDraft, plaidLinkTok
                     [BANK_INFO_STEP_KEYS.ACCOUNT_NUMBER]: '',
                     [BANK_INFO_STEP_KEYS.PLAID_MASK]: '',
-                    [BANK_INFO_STEP_KEYS.IS_SAVINGS]: '',
+                    [BANK_INFO_STEP_KEYS.IS_SAVINGS]: false,
```
why is this change needed?

mateuuszzzzz replied:
Before migration we had incorrect type value here. After migration `typecheck` throws type mismatch so it had to be changed. Since empty string is falsy I assume using `false` boolean value is fine here.

mountiny commented on src/types/form/ReimbursementAccountForm.ts:
```diff
@@ -21,6 +21,7 @@ const INPUT_IDS = {
         CITY: 'requestorAddressCity',
         STATE: 'requestorAddressState',
         ZIP_CODE: 'requestorAddressZipCode',
+        IS_ONFIDO_SETUP_COMPLETE: 'isOnfidoSetupComplete',
```
Why do we need this new input id here?

mateuuszzzzz replied:
It turned out we use it here https://github.com/software-mansion-labs/expensify-app-fork/blob/401bb219c74fb615704895554b6876eb25601a11/src/pages/ReimbursementAccount/RequestorOnfidoStep.js#L49 and missed it before

mountiny reviewed (commented):
@mateuuszzzzz thanks! Have you tested the full flow in the bank account and added the testing bank account succefully? Then also removed it? 

Seems like there is couple of changes different from the JS so curious why they needed and hence I think we need to be thorough with testing here

mateuuszzzzz commented:
> @mateuuszzzzz thanks! Have you tested the full flow in the bank account and added the testing bank account succefully? Then also removed it?
> 
> Seems like there is couple of changes different from the JS so curious why they needed and hence I think we need to be thorough with testing here

@mountiny  In general, I have a problem with verification via `Onfido` without a social security number, but this is not related to this PR. Similarly for `Plaid` even if the first steps are successful the verification step cannot be passed. Is there any example data that is valid and complete? All examples on open `#expensify-open-source` seem to be outdated or incomplete.

kubabutkiewicz commented:
![image](https://github.com/Expensify/App/assets/25230417/42941523-6988-450e-aa38-2b0a2b2e6cdb)
@mateuuszzzzz you can use those values to pass verification

mateuuszzzzz committed:
Remove redundant newline

mateuuszzzzz committed:
Merge branch 'main' into ts/migrate-reimbursement-account

mountiny reviewed (approved):
Thanks!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8329850329) to staging by https://github.com/mountiny in version: 1.4.54-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8351233131) to production by https://github.com/AndrewGable in version: 1.4.54-4 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
